### ISSUE TITLE
incorrect control code displayed in Windows console.

### ISSUE THREAD

**Issue by wenbingl** [google/python-fire#186]
when I run the basic example code. there are some control code displayed as output text. I guess it probably some color effects? 

How do I avoid it?

[1mFLAGS[0m
    --minputs=[4mMINPUTS[0m
    --moutputs=[4mMOUTPUTS[0m

[1mNOTES[0m
    You can also use flags syntax for POSITIONAL ARGUMENTS

**Comment by dbieber** [comment]
If you set the environment variable ANSI_COLORS_DISABLED=1 that should disable the coloring and your output should look normal.

**Comment by wenbingl** [comment]
great, it works, Thanks!
But is it an issue of  the other package used by Fire, so there is no plan to fix here?

**Comment by dbieber** [comment]
Yes, I would like to fix this. I will mark it as a bug.
Can you provide additional details about the system you're using?

What version of windows is it?
Is it command prompt or power shell?
Anything else I should know? (I'm not too familiar with windows development.)

**Comment by dbieber** [comment]
https://pypi.org/project/termcolor/ says that windows only supports "reverse" of the text attributes,
but we use bold and underline as well in Python Fire.

**Comment by wenbingl** [comment]
> Yes, I would like to fix this. I will mark it as a bug.
> Can you provide additional details about the system you're using?
> 
> What version of windows is it?
> Is it command prompt or power shell?
> Anything else I should know? (I'm not too familiar with windows development.)

I am running python code on Windows 10 **Version 1903 (OS Build 18362.295)**, But I don't think this Windows version matters, because I never saw a Windows console app can print the text in bold or with underline since the early Windows like 95, XP...

And, both the cmd window and powershell has the same issue. not sure if it is OK on some terminal tool porting from Linux, like mintty.

**Comment by dbieber** [comment]
Thanks for the info.

**Comment by fedorbirjukov** [comment]
Use platform.system() == 'Windows'?

**Comment by fedorbirjukov** [comment]
termcolor works in PowerShell on Windows 2016.

```
PS> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14393  3053
```

![termcolor-powershell](https://user-images.githubusercontent.com/1055607/65411311-d5c52c00-ddec-11e9-9f12-4b72037bd51e.png)

**Comment by wenbingl** [comment]
Pretty interesting,  it doesn't show the color effect, - in the error message of pythom, but the termcolor result looks bad in my machine.
![Annotation 2019-09-25 183202](https://user-images.githubusercontent.com/10278425/65651278-d6afb680-dfc2-11e9-9be5-4daa85d110d2.png)

**Comment by fedorbirjukov** [comment]
Ok, maybe it works only in virtualenv.

**Comment by truekonrads** [comment]
Quick hack - add this to the beginning of your script:
```
import os

if os.name=='nt':
    os.environ['ANSI_COLORS_DISABLED']="1"
```

**Comment by juacobarcelo** [comment]
In fire/formatting.py termcolor is used to print colored characters to console.

In Windows 10 console color codes are displayed instead of colored text.

It seems that is needed to import and initialize along with termcolor, colorama.

**Comment by jaredtrog** [comment]
@juacobarcelo Are you working on this one? I see there is a PR out but it may be worth additional investigation since not only the Powershell console will print colors but with the release of [Windows Terminal (Preview)](https://github.com/microsoft/terminal) there is fairly robust support for colors and other fancy formatting bits.

I use Windows (10, Server 2008 - 2019) with Fire somewhat frequently so I'd be up for taking a crack at this one.

Tagging @dbieber in case he is aware of someone else working it.

**Comment by dbieber** [comment]
@joejoevictor has looked into it some -- not sure if we have a solution on the horizon, maybe he can chime in with the latest. Neither of us has easy access to windows.

And thanks for raising the PR to my attention.

**Comment by jaredtrog** [comment]
I had a moment to poke at this...

@juacobarcelo was spot on with the colorama. Adding these two lines to formatting.py:22-23 immediately gave the desired results.

```
import colorama
colorama.init()
```

![image](https://user-images.githubusercontent.com/4381365/75587843-02850b80-5a3d-11ea-9a66-64a9287f459b.png)

A more robust solution might be to import colorama only if the OS is windows and colorama is installed, otherwise disable ANSI colors in the terminal. That would at least ensure the output didn't contain garble.

Additionally, colorama could be added as a fire dependency.

I'd have to dig a bit deeper to understand what required it but colorama was there after a clean install of python 3.8 on Windows 10.

Just some thoughts. I'll wait until others have a chance to jump in before attempting a solution.

**Comment by dbieber** [comment]
Awesome -- that's great to know.

**Comment by dbieber** [comment]
It looks like colorama doesn't handle underline.
@joejoevictor is planning to look into what prompt_toolkit does:
https://python-prompt-toolkit.readthedocs.io/en/master/pages/printing_text.html#ansi
It looks like it may have a solution that handles more of the formatting than colorama.

**Comment by jaredtrog** [comment]
@dbieber @joejoevictor I had another moment to poke at this one...

I checked out the colorama project to see why underline wasn't supported - it seems that it may be limited by fundamental support on legacy Windows platforms more than python libraries. [This](https://github.com/tartley/colorama/issues/38) issue has been out there since 2015 and has some interesting comments in it related to underline and reverse.

I did some quick tests with prompt_toolkit.ANSI and it seems it's only performing a similar function to termcolor where it processes ANSI escape sequences; the underlying escaped output doesn't seem to differ between libraries. It's quite possible that I've missed some functionality here that solves the problem but it wasn't apparent to me.

If that's the case then it may be worth considering colorama as an intermediary between no ANSI support and full support ANSI support on Windows in PowerShell and CMD consoles since it provides *almost* full support.

It's also worth noting that in my test environment (Windows 10) in CMD and PowerShell the underline components didn't error - they just weren't underlined. I didn't find that it detracted from the experience.

I fully appreciate that my experience doesn't represent everyone and it may be more desirable to have a long term solution in place. Just putting some ideas out there.

**Comment by dbieber** [comment]
Thanks for looking into it.
If I understand correctly, prompt_toolkit will just not underline the text, same as colorama.
If that's the case, your suggestion of the following (psuedocode) sounds good:

```
if os is windows:
  if colorama is installed:
    init colorama
  else:
    disable formatting
```

Is `os.name=='nt'` the right way to check for windows? Or is there something more robust?

**Comment by jaredtrog** [comment]
@dbieber  `os.name` certainly will work though heavy handed. As @fedorbirjukov pointed out, considering the console in use versus just the platform is important.

I took a look at the various methods used to detect the terminal and most were either incomplete or added more 3rd party libraries.

I came across an implementation of this very thing in the Chromium code base in a file called [setup_color.py](https://chromium.googlesource.com/chromium/tools/depot_tools.git/+/refs/heads/master/setup_color.py).

There are several items in here that relate to much older versions of Windows but it handles all the use cases we've discussed here and several edge cases that haven't. Also, this approach includes a function `enable_native_ansi` which works some *magic* in Windows 10 such that Underline is supported in CMD.

So the approach above works, we'd just init colorama through `setup_color.init()`. I've got some Windows Server 2k12r2 and some older Win10 machines that I'll test the solution against in addition to latest Win10, 2k16 and 2k19.

Thoughts on the method of initializing colorama?

**Comment by jaredtrog** [comment]
@dbieber Ok, completed some testing.

If fire only uses colorama without the extra work `setup_color.py` includes then output does not include any control code. Unsupported ANSI code is not displayed. For example, what should be underlined text in OPTIONAL output is not underlined. The output is human readable.

If fire imports and initializes colorama with the extra work `setup_color.py` provides then CMD and PowerShell correctly display ANSI code, including underlines, by enabling features of the respective console on Windows 10, Server 2016 and Server 2019.

Is it worth the extra code `setup_color.py` adds to the project to get more consistent output on modern Windows versions?

The solution I'm testing right now looks like this:
```
# formatting.py
try:
  import colorama
  HAS_COLORAMA = True
except ImportError:
  HAS_COLORAMA = False
  
if sys.platform.startswith('win'):
  if HAS_COLORAMA:
    colorama.init()
    # or
    # from setup_color import init
    # init()
  else:
    os.environ['ANSI_COLORS_DISABLED'] = "1"
```

The python docs show this method of detecting the platform as the recommended idiom to use. https://docs.python.org/3/library/sys.html#sys.platform.

**Comment by dbieber** [comment]
I haven't looked at setup_color.py in detail yet. Does it have dependencies?
If not and it's just a single Python file handling interesting edge cases for different systems, then it seems like it would be worthwhile to include it or a pared down version of it.

**Comment by jaredtrog** [comment]
No includes outside of core aside from colorama.

```
import ctypes
import os
import platform
import subprocess
import sys
from third_party import colorama # this would be changed in the PR
```

I'll put together a pull request with the proposed changes for further review and discussion.

**Comment by dbieber** [comment]
Awesome, sounds good. And maybe this is already what you planned, but let's not reuse the setup_color.py module from chromium directly, but rather reimplement as needed for fire.

**Comment by dbieber** [comment]
Looks great. Comments inline (and repeated here):
- I don't fully understand the criteria for setting `wrap` or if it should ever be set at all. (I assume that turns on some form of line wrapping)
- I think we should drop the print statement -- wdyt?

**Comment by dbieber** [comment]
Fixed in #232 
Will go out with the next release
